diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt

```diff
@@ -0,0 +1,183 @@
+import org.cqfn.diktat.ruleset.utils.getAllLeafsWithSpecificType
+import org.cqfn.diktat.ruleset.utils.isBeginByNewline
+import org.cqfn.diktat.ruleset.utils.isEol
+import org.cqfn.diktat.ruleset.utils.isFollowedByNewline
+import org.cqfn.diktat.ruleset.utils.isSingleLineIfElse
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.com.intellij.psi.tree.TokenSet
+import org.jetbrains.kotlin.psi.psiUtil.parents
+
+/**
+ * Rule that checks line break styles.
+ * 1. Prohibits usage of semicolons at the end of line
+ * 2. Checks that some operators are followed by newline, while others are prepended by it
+ * 3. Statements that follow `!!` behave in the same way
+ * 4. Forces functional style of chained dot call expressions with exception
+ */
+@Suppress("ForbiddenComment")
+class NewlinesRule : Rule("newlines") {
+    companion object {
+        // fixme: these token sets can be not full, need to add new once as corresponding cases are discovered.
+        // error is raised if these operators are prepended by newline
+        private val lineBreakAfterOperators = TokenSet.create(ANDAND, OROR, PLUS, PLUSEQ, MINUS, MINUSEQ, MUL, MULTEQ, DIV, DIVEQ)
+
+        // error is raised if these operators are followed by newline
+        private val lineBreakBeforeOperators = TokenSet.create(DOT, SAFE_ACCESS, ELVIS, COLONCOLON)
+
+        private val expressionTypes = TokenSet.create(DOT_QUALIFIED_EXPRESSION, SAFE_ACCESS_EXPRESSION, CALLABLE_REFERENCE_EXPRESSION, BINARY_EXPRESSION)
+        private val chainExpressionTypes = TokenSet.create(DOT_QUALIFIED_EXPRESSION, SAFE_ACCESS_EXPRESSION)
+    }
+
+    private lateinit var configRules: List<RulesConfig>
+    private lateinit var emitWarn: ((offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit)
+    private var isFixMode: Boolean = false
+
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       params: KtLint.Params,
+                       emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        configRules = params.getDiktatConfigRules()
+        isFixMode = autoCorrect
+        emitWarn = emit
+
+        if (node.elementType == SEMICOLON) {
+            handleSemicolon(node)
+        } else if (node.elementType in lineBreakAfterOperators) {
+            handleOperatorWithLineBreakAfter(node)
+        } else if (node.elementType in lineBreakBeforeOperators && !node.isDotFromPackageOrImport()) {
+            handleOperatorWithLineBreakBefore(node)
+        }
+    }
+
+    private fun handleSemicolon(node: ASTNode) {
+        if (node.isEol() && node.treeParent.elementType != ENUM_ENTRY) {
+            // semicolon at the end of line which is not part of enum members declarations
+            REDUNDANT_SEMICOLON.warnAndFix(configRules, emitWarn, isFixMode, node.extractLineOfText(), node.startOffset) {
+                node.treeParent.removeChild(node)
```

Comment:
treeParent always exists?

Reply:
Even if file consists of single semicolon, it's `treeParent` will be file itself, so should be safe.